MathlibTest/ring.lean

```diff
@@ -84,6 +84,29 @@ example (x : ℝ) (hx : x ≠ 0) :
   field_simp
   ring
 
+abbrev myId (a : ℤ) : ℤ := a
+
+/-
+Test that when `ring_nf` normalizes multiple expressions which contain a particular atom, it uses a
+form for that atom which is consistent between expressions.
+
+Note: This test is useless unless done at `=ₛ` (the level of syntactic equality); doing this test at
+`=` (the level of reducible equality) would not catch the erroneous old behaviour (normalizing one
+expression to `myId x * 2` and the other to `x * 2`).  But to get syntactic equality, we need to
+make the typeclass arguments agree, which requires some handholding in typeclass inference ... which
+is the reason for all the locally deleted instances here.
+-/
+attribute [-instance] instOfNat instNatCastInt Int.instSemiring Int.instOrderedCommRing
+  Int.instOrderedRing Int.instLinearOrderedCommRing Int.instMul NonUnitalNonAssocRing.toMul
+  NonUnitalNonAssocCommSemiring.toNonUnitalNonAssocSemiring
+  NonUnitalNonAssocRing.toNonUnitalNonAssocSemiring in
+example (x : ℤ) (R : ℤ → ℤ → Prop) : True := by
+  have : R (myId x + x) (x + myId x) := by
+    ring_nf
+    guard_target =ₛ R (myId x * 2) (myId x * 2)
```

Comment:
Here's an idea: after running `ring_nf`, do `rw [show myId = myOpaqueId from rfl]`, then you don't need the `attribute [-instance]`s

Reply:
I took a guess at interpreting your suggestion here: I added
```lean
opaque myOpaqueId (a : ℤ) : ℤ := a
```
before the test and then adding `rw [show myId = myOpaqueId from rfl]` below the `guard_target`.  This fails because, well, `myOpaqueId` is opaque.

I guess you meant something else.  Please push it.

Reply:
Thanks, that's in the distribution of things I tried to mean :). I pushed a working version.

Reply:
What's tricky about your proposed test is that now it depends on `rw`'s notion of equality, as well as `guard_target`'s.  And I don't find `rw`'s notion of equality particularly principled or well-documented: it's stricter than reducible defeq, but it's still not literal equality of expressions -- it's something like "reducible equality, together with the same head-symbol, after unfolding abbreviations".

Are you particularly concerned about the robustness of the original test?  I'm worried that the changed test is vulnerable to changes in `rw` down the line.

Reply:
```lean
import Mathlib
variable (P : ℝ → Prop) (f : ℝ → ℝ) (x : ℝ) (Q : Prop)

-- `rw` FAILS
-- `(fun t ↦ P t) (f x)` and `P (f x)` reducibly defeq with different head
example (h : (fun t ↦ P t) (f x) = Q) : P (f x) := by rw [h]

-- `rw` WORKS
-- `P ((fun t ↦ f t) x)` and `P (f x)` reducibly defeq with same head
example (h : P ((fun t ↦ f t) x) = Q) : P (f x) := by rw [h]; sorry
```

Reply:
> Are you particularly concerned about the robustness of the original test?

Yes, I worry that the list of instances to ignore in the original test is going to be annoying to maintain.

I agree that `rw` isn't particularly well specified; which is why I entrusted it to replaces the simplest expression possible, one constant for another with no binders or variables. Perhaps using `rewrite` would make it a little safer.

Perhaps it's worth getting @digama0 or @kmill's opinion here on whether there is a middle ground between the two approaches, or if `rw` is trustworthy enough?

Reply:
How about this then?  It doesn't depend on `rw`, and it tests "pretty-printer equality", which is arguably more principled than "`rw` equality".
```lean
/--
error: unsolved goals
x : ℤ
R : ℤ → ℤ → Prop
h : R (myId x * 2) (myId x * 2)
⊢ True
-/
#guard_msgs in
example (x : ℤ) (R : ℤ → ℤ → Prop) : True := by
  have h : R (myId x + x) (x + myId x) := sorry
  ring_nf at h
```

Reply:
(Not saying you should use it here, but `trace_state` is a tactic that prints the current tactic state. It's more useful in tests that go through multiple states. Relying on the 'unsolved goals' error seems fine to me.)

Reply:
That looks good to me too, though I like the `trace_state` suggestion.

Reply:
Another tip, set `set_option pp.mvars false` to make the test be insensitive to the metavariable numbering.

Reply:
Thanks for the hints Kyle!